File: 0712-minimum-ascii-delete-sum-for-two-strings/0712-minimum-ascii-delete-sum-for-two-strings.cpp
class Solution {

public:
    int minimumDeleteSum(string s1, string s2) {
        int p = s1.length(), q = s2.length();
        vector<vector<int>> dp;
        dp.resize(p + 1, vector<int>(q + 1, 1000000));


        dp[0][0] = 0;

        for (int i = 0; i <= p; i++)
        {
            for (int j = 0; j <= q; j++)
            {
                // cout << i << ' ' << j << ' ' << dp[i][j] << '\n';
                if (i != p && j != q && s1[i] == s2[j])
                {
                    // cout << ' ' << i + 1 << ' ' << j + 1<< ' ' << dp[i + 1][j + 1] << ' ' << dp[i][j]  << '\n';
                    dp[i + 1][j + 1] = min(dp[i + 1][j + 1], dp[i][j]);
                }

                if (i != p) {
                    dp[i + 1][j] = min(dp[i + 1][j], dp[i][j] + s1[i]);
                    // cout << '\t' << i + 1 << ' ' << j << ' ' << dp[i + 1][j] << ' ' << dp[i][j]  << '\n';
                }

                if (j != q) {
                    dp[i][j + 1] = min(dp[i][j + 1], dp[i][j] + s2[j]);
                    // cout << "\t\t\t" << i << ' ' << j + 1 << ' ' << dp[i][j + 1] << ' ' << dp[i][j]  << '\n';
                }
                if (i != p && j != q) {
                    dp[i + 1][j + 1] = min(dp[i + 1][j + 1], dp[i][j] + s1[i] + s2[j]);
                    // cout << "\t\t" << i + 1 << ' ' << j + 1 << ' ' << dp[i + 1][j + 1] << ' ' << dp[i][j] + s1[i] + s2[j] << '\n';
                }
            }
        }
        return dp[p][q];
    }
};

File: 0712-minimum-ascii-delete-sum-for-two-strings/0712-minimum-ascii-delete-sum-for-two-strings.java
class Solution {

    private int getAsciiSum(String s)
    {
        int ret = 0;
        for (char c : s.toCharArray())
        {
            ret += c;
        }
        return ret;
    }
    public int minimumDeleteSum(String s1, String s2) {
        int[][] dp = new int[1001][1001];
        int s1l = s1.length();
        int s2l = s2.length();


        for (int i = 0; i <= s1l; i++)
        {
            for (int j = 0; j <= s2l; j++)
            {
                if (i != s1l && j != s2l && s1.charAt(i) == s2.charAt(j))
                {
                    dp[i + 1][j + 1] = Math.max(dp[i + 1][j + 1], dp[i][j] + (int) s1.charAt(i) + (int) s2.charAt(j));
                }
                else
                {
                    if (j != s2l)
                    {
                        dp[i ][j + 1] = Math.max(dp[i][j + 1], dp[i][j]);   
                    }
                    if (i != s1l)
                    {
                        dp[i + 1][j] = Math.max(dp[i + 1][j], dp[i][j]);   
                    }
                }
            }
        }
        return getAsciiSum(s1) + getAsciiSum(s2) - dp[s1l][s2l];
    }
}